### ISSUE TITLE
Correlating the codebase with paper 

### ISSUE THREAD

**Issue by ScottHoang** [NVlabs/ESPACE#1]
Dear authors, 
Thank you so much for releasing the source code; I really appreciate all the hard work that went into it. I have some questions related to its implementation I hope you can enlighten me. 

1. During the calibration phase,  I assume `model_modifier` loops through every module within a model and replaced it with `TensorStatisticsEstimator,` which is an `nn.Identity` that calculates autocorrelation on the side. Assuming I want to calibrate every `nn.Linear` within my model, wouldn't this destroy my model's intermediate activations, leading to sub-optimal calibration of input correlation?
2. `weight_identity` is referenced throughout the code base to be merged with `mse` and `nmse` to generate the `go_` versions. But I don't see where it is being calculated. Is this a pre-computation step done somewhere else?
3. In the paper, you described choosing the correct criteria per layer based on its sensitivity to wikitext ppl. Is this step done one layer per evaluation, or performed progressively one layer at a time (i.e by keeping the previously calibrated layer activated)? 

Thank you so much for your work. I hope to hear back from you.

**Comment by csakr** [comment]
Hi,

Thanks for sending these questions.

For 1) what you want to do is to send your activations and weights though an `nn.Identity`  - activations are simply intermediate results. For weights, if you are using `nn.Linear` then you will need to re-write it as a module that explicitly lets you access the weight tensor. For instance, checkout ColumnParallelLinear and RowParallelLinear layers from Megatron. Regarding the destruction of activations, this will not happen since the output of the module will be forwarded as is. We use `TensorStatisticsEstimator` only to probe statistics related to activations, and do not modify the activations themselves.

For 2) in the `go_` versions we need to multiply activation autocorrelation matrices with weight autocorrelation matrices. Note that the latter is simply WW^T (see paper). So, we just set the `weight_identity` to become a `TensorStatisticsEstimator` and it will return what we want. At the end we multiply the autocorrelation sitting in the activation estimator to that of the weight, and we get the desired result.

For 3) what we did in the paper specifically is one layer per evaluation, and then we progressively apply the projection for lest destructive layers found in the first step (this is detailed in section 4 and the appendix of the paper I believe). Note that we did this because it makes sense from the perspective of keeping accuracy degradation low. You are welcome to investigate other strategies.
